### ISSUE TITLE
DataCollatorForLanguageModeling ignores manual -100 labels for loss masking

### ISSUE THREAD

**Issue by kabachuha** [huggingface/transformers#26357]
### System Info

transformers version https://github.com/huggingface/transformers/commit/5936c8c57ccb2bda3b3f28856a7ef992c5c9f451
Kubuntu 22.04 LTS
NVIDIA 4090

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Info from https://github.com/oobabooga/text-generation-webui/issues/4031

Consider oobabooga's training script

https://github.com/oobabooga/text-generation-webui/blob/ee7bf49804bfb53775424d4c12294445497b610f/modules/training.py#L341-L358

It uses -100 labels in attempt to mask the loss for the parts we don't need to train or, like when we use `## RESPONSE` masking for better training of Alpaca lora, excluding retraining on the beginning part of the dialogue.

---

To reproduce just put prints to DataCollatorForLanguageModeling's `torch_call` function:

```python
labels = batch["input_ids"].clone()
print('INPUT LABELS')
print(labels)
if self.tokenizer.pad_token_id is not None:
    labels[labels == self.tokenizer.pad_token_id] = -100
print('OVERWRITTEN LABELS')
batch["labels"] = labels
print(batch["labels"])
```

### Expected behavior

The collator should not overwrite the manual set masks for loss calculation, as it creates confusion when the trainers set the labels to -100 only to find out it doesn't actually mask the loss. This is extremely non-obvious (need to look up the very base transformers code) and has already cost me (and countless clueless users) days of wasted high-end GPU time

**Comment by LysandreJik** [comment]
Hey @kabachuha, this seems like a reasonable feature request! Would you like to open a PR to offer your fix?

**Comment by kabachuha** [comment]
Thanks for your interest! Sure, I can, it is quite short.
